dodiscover/context_builder.py

```diff
@@ -0,0 +1,214 @@
+
+        Returns
+        -------
+        ContextBuilder
+            The builder instance
+        """
+        self._state_variables = state_variables
+        return self
+
+    def state_variable(self, name: str, var: Any) -> "ContextBuilder":
+        """Add a state variable.
+
+        Called by an algorithm to persist data objects that
+        are used in intermediate steps.
+
+        Parameters
+        ----------
+        name : str
+            The name of the state variable.
+        var : any
+            Any state variable.
+        """
+        if self._state_variables is None:
+            self._state_variables = dict()
+
+        self._state_variables[name] = var
+        return self
+
+    def build(self) -> Context:
+        """Build the Context object.
+
+        Returns
+        -------
+        Context
+            The populated Context object
+        """
+        if self._observed_variables is None:
+            raise ValueError("Could not infer variables from data or given arguments.")
+        return Context(
+            init_graph=self._interpolate_graph(),
+            included_edges=self._interpolate_included_edges(),
+            excluded_edges=self._interpolate_excluded_edges(),
+            variables=self._observed_variables,
+            latents=self._latent_variables or set(),
+            state_variables=self._state_variables or dict(),
+        )
+
+    def _interpolate_graph(self) -> nx.Graph:
+        if self._observed_variables is None:
+            raise ValueError("Must set variables() before building Context.")
+
+        variables = self._observed_variables or set()
+        graph = self._graph
+        # initialize the starting graph
+        if graph is None:
+            return nx.complete_graph(variables, create_using=nx.Graph)
+        else:
+            if set(graph.nodes) != set(variables):
+                raise ValueError(
+                    f"The nodes within the initial graph, {graph.nodes}, "
+                    f"do not match the nodes in the passed in data, {variables}."
+                )
+            return graph
+
+    def _interpolate_included_edges(self) -> Union[nx.Graph, nx.DiGraph]:
+        # initialize set of fixed and included edges
+        return self._included_edges or nx.empty_graph(
+            self._observed_variables, create_using=nx.Graph
+        )
+
+    def _interpolate_excluded_edges(self) -> Union[nx.Graph, nx.DiGraph]:
+        # initialize set of fixed and included edges
+        return self._excluded_edges or nx.empty_graph(
+            self._observed_variables, create_using=nx.Graph
+        )
+
+
+def make_context(context: Optional[Context] = None) -> ContextBuilder:
+    """Create a new ContextBuilder instance.
+
+    Returns
+    -------
+    ContextBuilder
+        The new ContextBuilder instance
+    """
+    result = ContextBuilder()
+    if context is not None:
+        result.graph(context.init_graph)
```

Comment:
Could we just make use of context.copy() here to get all necessary info? 

Just thinking that these LOC might become a pain to maintain every time we add something to context.